File: framework/fit/java/fit-api/src/main/java/com/huawei/fitframework/broker/Genericable.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2023-2023. All rights reserved.
 */

package com.huawei.fitframework.broker;

import java.util.List;

/**
 * 表示可执行的泛服务对象。
 *
 * @author 季聿阶 j00559309
 * @since 2023-03-07
 */
public interface Genericable extends GenericableMetadata {
    /**
     * 获取服务的所有实现列表。
     *
     * @return 表示服务的所有实现列表的 {@link List}{@code <}{@link Fitable}{@code >}。
     */
    @Override
    List<Fitable> fitables();

    /**
     * 执行服务。
     *
     * @param context 表示调用上下文的 {@link InvocationContext}。
     * @param args 表示调用参数列表的 {@link Object}{@code []}。
     * @return 表示调用结果的 {@link Object}。
     */
    Object execute(InvocationContext context, Object[] args);
}

File: framework/fit/cpp/plugin/registry_listener/src/domain/genericable.cpp
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2021. All rights reserved.
 *
 * Description  : Provides implementation for base of genericable.
 * Author       : liangjishi 00298979
 * Date         : 2021/12/08
 */

#include <domain/genericable.hpp>

#include <registry_listener.hpp>
#include <repo_factory.hpp>
#include <util/singleton_utils.hpp>

using namespace Fit;
using namespace Fit::Registry::Listener;

Genericable::Genericable(const GenericableRepoPtr& repo, Fit::string id, Fit::string version)
    : repo_(repo), id_(std::move(id)), version_(std::move(version))
{
}

RegistryListenerPtr Genericable::GetRegistryListener() const
{
    GenericableRepoPtr repo = repo_.lock();
    return (repo == nullptr) ? nullptr : (repo->GetRegistryListener());
}

const Fit::string& Genericable::GetId() const
{
    return id_;
}

const Fit::string& Genericable::GetVersion() const
{
    return version_;
}

FitableRepoPtr Genericable::GetFitables()
{
    return SingletonUtils::Get<FitableRepo>(fitables_, mutex_, [&]() {
        return GetRegistryListener()->GetRepoFactory()->CreateFitableRepo(shared_from_this());
    });
}

int32_t Genericable::Compare(const GenericablePtr& another) const
{
    return Compare(another->GetId(), another->GetVersion());
}

int32_t Genericable::Compare(const string& id, const string& version) const
{
    int32_t ret = GetId().compare(id);
    if (ret == 0) {
        ret = GetVersion().compare(version);
    }
    return ret;
}
